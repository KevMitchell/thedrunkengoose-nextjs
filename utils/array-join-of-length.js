export function joinString (arr = [], length = 1000, delimeter = '') {
  if (!arr.length) return ''

  let result = ''
  let currentLineLength = 0
  arr.forEach((item, index) => {
    if (index >= arr.length - 1) {
      result = result.concat(item)
    } else {
      currentLineLength += String(item).length
      const nextItemLength = String(arr[index + 1]).length + delimeter.length
      if (currentLineLength + nextItemLength > length) {
        result = result.concat(item).concat('\n')
        currentLineLength = 0
      } else {
        result = result.concat(item).concat(delimeter)
      }
    }
  })
  return result
}

export function joinJsx (arr = [], length = 1000, delimeter = '') {
  const createDiv = (elements, key) =>
    <div key={key}>{elements.reduce((a, c) => a.concat(delimeter).concat(c), '').replace(delimeter, '')}</div>

  const splitIntoLines = () => {
    let lineStart = 0
    let currentLineLength = 0
    const result = []
    for (let i = 0; i < arr.length; ++i) {
      if (i === arr.length - 1) {
        result.push(arr.slice(lineStart))
      } else {
        currentLineLength += String(arr[i]).length
        const nextItemLength = String(arr[i + 1]).length + delimeter.length
        if (currentLineLength + nextItemLength > length) {
          result.push(arr.slice(lineStart, i + 1))
          currentLineLength = 0
          lineStart = i + 1
        }
      }
    }

    return result
  }

  return (
    <>
      {splitIntoLines().map(createDiv)}
    </>
  )
}
