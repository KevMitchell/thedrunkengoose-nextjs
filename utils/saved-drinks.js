export function toggleSaved (name, saved) {
  if (saved) {
    localStorage.setItem(name, 'saved')
  } else {
    localStorage.removeItem(name)
  }
}

export function getSaved (name) {
  return localStorage.getItem(name) === 'saved'
}
