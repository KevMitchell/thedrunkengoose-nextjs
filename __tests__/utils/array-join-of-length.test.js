import { render } from '@testing-library/react'
import { joinString, joinJsx } from '../../utils/array-join-of-length'

describe('joinString', () => {
  test('returns an empty string for an empty array', () => {
    expect(joinString([])).toEqual('')
  })

  describe('small array', () => {
    const smallArray = ['a', 'b']
    test('returns a joined string if array items fit onto one line', () => {
      expect(joinString(smallArray, 10)).toEqual('ab')
    })

    test('uses delimeters if provided', () => {
      expect(joinString(smallArray, 10, '-')).toEqual('a-b')
    })
  })

  describe('large array', () => {
    const largeArray = ['aaa', 'bbbb']
    test('returns a joined string where array items do not fit onto one line', () => {
      expect(joinString(largeArray, 3)).toEqual('aaa\nbbbb')
    })

    test('uses delimeters if provided', () => {
      expect(joinString(largeArray, 10, '-')).toEqual('aaa-bbbb')
    })

    test('doesnt use delimeters with newlines', () => {
      expect(joinString(largeArray, 3, '-')).toEqual('aaa\nbbbb')
    })
  })
})

describe('joinJsx', () => {
  test('returns an empty element for an empty array', () => {
    const { container } = render(joinJsx([]))
    expect(container).toBeEmptyDOMElement()
  })

  describe('small array', () => {
    const smallArray = ['a', 'b']

    test('returns a single element if array items fit onto one line', () => {
      const { container } = render(joinJsx(smallArray, 10))
      expect(container).toMatchSnapshot()
    })

    test('uses delimeters if provided', () => {
      const { container } = render(joinJsx(smallArray, 10, '-'))
      expect(container).toMatchSnapshot()
    })
  })

  describe('large array', () => {
    const largeArray = ['aaa', 'bbbb', 'ccccccccc', 'ddd']

    test('returns multiple elements where array items do not fit onto one line', () => {
      expect(joinJsx(largeArray, 8)).toMatchSnapshot()
    })

    test('uses delimeters if provided', () => {
      expect(joinJsx(largeArray, 8, '-')).toMatchSnapshot()
    })
  })
})
