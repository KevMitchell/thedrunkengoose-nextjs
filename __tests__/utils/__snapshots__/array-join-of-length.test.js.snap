// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`joinJsx large array returns multiple elements where array items do not fit onto one line 1`] = `
<React.Fragment>
  <div>
    aaabbbb
  </div>
  <div>
    ccccccccc
  </div>
  <div>
    ddd
  </div>
</React.Fragment>
`;

exports[`joinJsx large array uses delimeters if provided 1`] = `
<React.Fragment>
  <div>
    aaa-bbbb
  </div>
  <div>
    ccccccccc
  </div>
  <div>
    ddd
  </div>
</React.Fragment>
`;

exports[`joinJsx small array returns a single element if array items fit onto one line 1`] = `
<div>
  <div>
    ab
  </div>
</div>
`;

exports[`joinJsx small array uses delimeters if provided 1`] = `
<div>
  <div>
    a-b
  </div>
</div>
`;
